gridstatus/miso.py

```diff
@@ -679,8 +679,586 @@ def _get_generation_outages_data(
+                "BP3",
+                "PC3",
+                "BP4",
+                "PC4",
+                "OVERRIDE",
+                "REASON",
+            ]
+        ]
+
+    @support_date_range(frequency="DAY_START")
+    def get_reserve_product_binding_constraints_day_ahead_hourly(
+        self,
+        date: str | pd.Timestamp,
+        end: str | pd.Timestamp = None,
+        verbose: bool = False,
+    ):
+        query_date = date - pd.Timedelta("1D")
+        url = f"https://docs.misoenergy.org/marketreports/{query_date.strftime('%Y%m%d')}_da_rpe.xls"
+        logger.info(
+            f"Downloading day-ahead reserve product binding constraints data from {url}",
+        )
+
+        excel_file = pd.ExcelFile(url)
+        market_date, publish_date = self._get_constraint_header_dates_from_excel(
+            excel_file,
+        )
+        data = pd.read_excel(excel_file, skiprows=3)
+
+        # NOTE(kladar): The last row is a text disclaimer, and there is a leading space
+        # in the column names, so we clean it all up.
+        data = data.iloc[:-1]
+        data["Interval End"] = market_date + pd.to_timedelta(
+            data[
+                "Hour of Occurence"
+            ],  # NOTE(kladar): sic, this is a persistent typo in the header from MISO
+            unit="h",
+        )
+        data["Interval Start"] = data["Interval End"] - pd.Timedelta(hours=1)
+
+        return data[
+            [
+                "Interval Start",
+                "Interval End",
+                "Constraint Name",
+                "Shadow Price",
+                "Constraint Description",
+            ]
+        ]
+
+    @support_date_range(frequency="DAY_START")
+    def get_binding_constraints_real_time_5_min(
+        self,
+        date: str | pd.Timestamp,
+        end: str | pd.Timestamp = None,
+        verbose: bool = False,
+    ):
+        query_date = date + pd.Timedelta("1D")
+        url = f"https://docs.misoenergy.org/marketreports/{query_date.strftime('%Y%m%d')}_rt_bc.xls"
+        logger.info(f"Downloading real-time binding constraints data from {url}")
+
+        excel_file = pd.ExcelFile(url)
+        market_date, publish_date = self._get_constraint_header_dates_from_excel(
+            excel_file,
+        )
+        data = pd.read_excel(
+            excel_file,
+            skiprows=3,
+            dtype={
+                "Constraint Description": object,
+                "Preliminary Shadow Price": float,
+                "BP1": float,
+                "PC1": float,
+                "BP2": float,
+                "PC2": float,
+            },
+        )
+
+        data["Interval End"] = pd.to_datetime(
+            market_date.strftime("%Y-%m-%d")
+            + " "
+            + data[
+                "Hour of  Occurrence"
+            ],  # NOTE(kladar): sic, there are two spaces between "Hour of" and "Occurrence"
+        ).dt.tz_localize(
+            self.default_timezone,
+        )
```

Comment:
When calling this method, the data starts 5 minutes before the passed date and ends 5 minutes before the end of the date. Is it possible this line should be for the `Interval Start`?

<img width="475" alt="Screenshot 2024-11-12 at 15 36 13" src="https://github.com/user-attachments/assets/631d393b-4da6-47e9-a85e-748f2c268eb5">

Reply:
I had a similar thought, but Connor confirmed that these are indeed interval ending. https://github.com/gridstatus/gridstatus/pull/476#discussion_r1837435304

Since they are individual date files (e.g. `...2024-11-10_rt_bc.xls`), unless we want to get fancy with multiple file pulls+data stitching for a single day, it seems it will be off by an interval due to the MISO convention.

Reply:
Alright, thanks. It's very weird to me that MISO publishes the last interval of the previous day with the next day's data.